src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -647,7 +647,10 @@ class RsFnInferenceContext(
         when {
             blockExpr.isTry -> inferTryBlockExprType(blockExpr, expected)
             blockExpr.isAsync -> inferAsyncBlockExprType(blockExpr)
-            else -> blockExpr.block.inferType(expected)
+            else -> {
+                val type = blockExpr.block.inferType(expected)
+                inferLabeledExprType(blockExpr, type, true)
```

Comment:
I didn't investigate deeply how  make type inference for such case properly - just generalize code of loop expr type inference